pkg/sdn/plugin/controller.go

```diff
@@ -330,6 +337,155 @@ func (plugin *OsdnNode) SetupSDN(localSubnetCIDR, clusterNetworkCIDR, servicesNe
+			return
+		}
+		policy := obj.(*osapi.EgressNetworkPolicy)
+
+		vnid, err := plugin.vnids.GetVNID(policy.Namespace)
+		if err != nil {
+			glog.Warningf("Could not find namespace %q for EgressNetworkPolicy: %v", err)
+			continue
+		}
+
+		policies := plugin.egressPolicies[vnid]
+		for i, oldPolicy := range policies {
+			if oldPolicy.UID == policy.UID {
+				policies = append(policies[:i], policies[i+1:]...)
+				break
+			}
+		}
+		if eventType != watch.Deleted && len(policy.Spec.Egress) > 0 {
+			policies = append(policies, policy)
+		}
+		plugin.egressPolicies[vnid] = policies
+
+		err = plugin.updateEgressNetworkPolicy(vnid)
+		if err != nil {
+			utilruntime.HandleError(err)
+			return
+		}
+	}
+}
+
+func (plugin *OsdnNode) UpdateEgressNetworkPolicyVNID(namespace string, oldVnid, newVnid uint32) error {
+	var policy *osapi.EgressNetworkPolicy
+
+	policies := plugin.egressPolicies[oldVnid]
+	for i, oldPolicy := range policies {
+		if oldPolicy.Namespace == namespace {
+			policy = oldPolicy
+			plugin.egressPolicies[oldVnid] = append(policies[:i], policies[i+1:]...)
+			err := plugin.updateEgressNetworkPolicy(oldVnid)
+			if err != nil {
+				return err
+			}
+			break
+		}
+	}
+
+	if policy != nil {
+		plugin.egressPolicies[newVnid] = append(plugin.egressPolicies[newVnid], policy)
+		err := plugin.updateEgressNetworkPolicy(newVnid)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func policyNames(policies []*osapi.EgressNetworkPolicy) string {
+	names := make([]string, len(policies))
+	for i, policy := range policies {
+		names[i] = policy.Namespace + ":" + policy.Name
+	}
+	return strings.Join(names, ", ")
+}
+
+func (plugin *OsdnNode) updateEgressNetworkPolicy(vnid uint32) error {
+	otx := ovs.NewTransaction(kexec.New(), BR)
+
+	policies := plugin.egressPolicies[vnid]
+	namespaces := plugin.vnids.GetNamespaces(vnid)
+	if len(policies) == 0 {
+		otx.DeleteFlows("table=9, reg0=%d", vnid)
+	} else if vnid == 0 {
+		glog.Errorf("EgressNetworkPolicy in global network namespace is not allowed (%s); ignoring", policyNames(policies))
+	} else if len(namespaces) > 1 {
+		glog.Errorf("EgressNetworkPolicy not allowed in shared NetNamespace (%s); dropping all traffic", strings.Join(namespaces, ", "))
+		otx.DeleteFlows("table=9, reg0=%d", vnid)
+		otx.AddFlow("table=9, reg0=%d, priority=1, actions=drop", vnid)
+	} else if len(policies) > 1 {
+		glog.Errorf("multiple EgressNetworkPolicies in same network namespace (%s) is not allowed; dropping all traffic", policyNames(policies))
+		otx.DeleteFlows("table=9, reg0=%d", vnid)
+		otx.AddFlow("table=9, reg0=%d, priority=1, actions=drop", vnid)
+	} else /* vnid != 0 && len(policies) == 1 */ {
+		// Temporarily drop all outgoing traffic, to avoid race conditions while modifying the other rules
+		otx.AddFlow("table=9, reg0=%d, cookie=1, priority=65535, actions=drop", vnid)
+		otx.DeleteFlows("table=9, reg0=%d, cookie=0/1", vnid)
+
```

Comment:
we do; the comment about dropping all traffic only applies to the AddFlow(); the DeleteFlows() then drops all existing flows for vnid (except for the "drop all traffic" flow which we just added with cookie=1 to mark it as being different from the the other flows)